### PR TITLE
Refactor targeting pack downloads, and framework version selection

### LINKED ISSUE #2792: RuntimeFrameworkVersion for 3.0 apps
Filing on behalf of @DamianEdwards @davidfowl 

It appears the RuntimeFrameworkVersion property is no longer honored in netcoreapp3.0 projects. If our recollection is correct, the plan was to use this property so customers could define the versions used by FrameworkReference. 

Open question: how to handle the difference between shared framework versions. After RTM, we should (in theory) always patch all in-box shared frameworks. But during per-release, however, these versions will rarely align.

### PR DESCRIPTION
**Pull request by dsplaisted**
This is currently WIP, I'm sending the PR now to get CI results.

This includes the following:

- Refactors FrameworkReference and targeting pack resolution so that all applicable targeting packs for the target framework will be downloaded.  This is required to support dotnet/cli#10666 (once NuGet/Home#7342 is implemented)
- Fixes #2792.  The following are the ways (in priority order) that the "runtime framework version" for a shared framework can be controlled:
    - `RuntimeFrameworkVersion` metadata on the `FrameworkReference` item
    - `RuntimeFrameworkVersion` MSBuild property (applies to all shared frameworks)
    - Then, use either the `LatestRuntimeFrameworkVersion` or the `DefaultRuntimeFrameworkVersion` of the `KnownFrameworkReference`, based on
        - The value (if set) of `TargetLatestRuntimePatch` metadata on the `FrameworkReference`
        - The `TargetLatestRuntimePatch` MSBuild property (which defaults to True if `SelfContained` is true, and False otherwise)

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveRuntimePackAssets.cs [hunk 1]
```diff
@@ -12,15 +12,25 @@ public class ResolveRuntimePackAssets : TaskBase
...
         public ITaskItem[] ResolvedRuntimePacks { get; set; }
 
+        public ITaskItem[] FrameworkReferences { get; set; } = Array.Empty<ITaskItem>();
+
         [Output]
         public ITaskItem[] RuntimePackAssets { get; set; }
 
         protected override void ExecuteCore()
         {
             List<TaskItem> runtimePackAssets = new List<TaskItem>();
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveRuntimePackAssets.cs]: suggestion: extract some methods or even classes. This method is too long. Too many nested layer

**Comment by dsplaisted** [comment]
@dotnet/dotnet-cli @nguerrera @wli3 These code changes are now ready for review.  I still need to look at adding more test coverage.

**Comment by nguerrera** [comment]
(Reading description, not code yet...)

Does this mean that there will have to be a targeting pack package produced for every servicing event, even if there is no API change?

**Comment by dsplaisted** [comment]
The `RuntimeFrameworkVersion` (either as metadata or a property) isn't used for the targeting pack version.  You can set the `TargetingPackVersion` metadata on a `FrameworkReference` to specify the targeting pack version (with these changes).

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs [hunk 2]
```diff
@@ -24,51 +31,71 @@ public class ResolveTargetingPackAssets : TaskBase
...
+                    }
                 }
             }
 
+            HashSet<string> frameworkReferenceNames = new HashSet<string>(FrameworkReferences.Select(fr => fr.ItemSpec), StringComparer.OrdinalIgnoreCase);
+            RuntimeFrameworksToRemove = RuntimeFrameworks.Where(rf => !frameworkReferenceNames.Contains(rf.GetMetadata(MetadataKeys.FrameworkName)))
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs]: just question, you need to remove RuntimeFrameworks?

Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs]: ResolveFrameworkReferences now resolves all the possible shared frameworks that could be referenced.  This is because eventually we will support [transitive shared framework references](https://github.com/dotnet/cli/issues/10666).  It won't be until we run ResolvePackageAssets that we know for sure which ones will be referenced, so we need to filter the targeting packs, runtime packs, and RuntimeFrameworks after that target runs.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                    var packTargetFramework = NuGetFramework.Parse(appHostPack.GetMetadata("TargetFramework"));
+                    return packTargetFramework.Framework.Equals(TargetFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) &&
+                        ResolveFrameworkReferences.NormalizeVersion(packTargetFramework.Version) ==
+                            ResolveFrameworkReferences.NormalizeVersion(new Version(TargetFrameworkVersion));
+                })
+                .ToList();
+
+            if (knownAppHostPacksForTargetFramework.Count > 1)
+            {
+                throw new InvalidOperationException("Multiple KnownAppHostPack items applied to the specified target framework, which is not supposed to happen");
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs]: I'm torn on this being a non-BuildErrorException because KnownAppHostPack could be modified by anyone. Unlikely though, and I guess this is just moved from elsewhere.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs [hunk 4]
```diff
@@ -55,167 +51,173 @@ public class ResolveFrameworkReferences : TaskBase
...
+                    targetingPackVersion = frameworkReference.GetMetadata("TargetingPackVersion");
+                }
+                if (string.IsNullOrEmpty(targetingPackVersion))
                 {
-                    if (!string.IsNullOrEmpty(knownFrameworkReference.AppHostPackNamePattern))
+                    targetingPackVersion = knownFrameworkReference.TargetingPackVersion;
+                }
+                targetingPack.SetMetadata(MetadataKeys.PackageVersion, targetingPackVersion);
+
+                targetingPack.SetMetadata(MetadataKeys.RelativePath, "");
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: Why do we need an explicit empty metadata value? Is this so that it can become "data" folder later, and also not the same place for all targeting packs all at once?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.TargetingPackResolution.targets [hunk 5]
```diff
@@ -38,6 +39,15 @@ Copyright (c) .NET Foundation. All rights reserved.
...
       -->
+      
+      <!-- TODO: Add this to BundledVersions in dotnet core-sdk, and remove it from here -->
+      <KnownAppHostPack Condition="'@(KnownAppHostPack)' == ''"
+                        Include="Microsoft.NETCore.App"
+                        TargetFramework="netcoreapp3.0"
+                        AppHostPackNamePattern="runtime.**RID**.Microsoft.NETCore.DotNetAppHost"
+                        AppHostPackVersion="3.0.0-preview-27218-01"
+                        AppHostRuntimeIdentifiers="freebsd-x64;linux-arm;linux-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
+                        />
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.TargetingPackResolution.targets]: Didn't you do this already?

**Comment by nguerrera** [comment]
Do we have a read on the performance cost of these new Resolve tasks? I am holding back on commenting on all the Linq and what not because I'm guessing N is small enough for it not to matter and for there to be lots of other allocations happening in msbuild nearby, but I am getting curious.

Overall looks good. I think @wli3 makes a good suggestion to try to extract some smaller methods.

**Comment by nguerrera** [comment]
I'm seeing this on a relatively recent build of dotnet new console
````
Target Performance Summary       
       14 ms  ResolveProjectReferences                   1 calls
       15 ms  _HandlePackageFileConflicts                1 calls
       20 ms  ResolveTargetingPacks                      1 calls
       21 ms  PrepareForBuild                            1 calls
       22 ms  ResolveAssemblyReferences                  1 calls
       22 ms  FindReferenceAssembliesForReferences       1 calls
       89 ms  ResolveFrameworkReferences                 1 calls

Task Performance Summary:
         8 ms  ResolvePackageAssets                       1 calls
       14 ms  ResolvePackageFileConflicts                1 calls
       16 ms  GetPackageDirectory                        2 calls
       22 ms  ResolveAssemblyReference                   1 calls
       88 ms  ResolveFrameworkReferences                 1 calls
```
I think we need to be more frugal in these tasks. I spent a lot of time optimizing ResolvePackageAssets. We have to be careful moving to something else that happens on incremental build for framework references.

We need to dig in. I am surprised that GetPackageDirectory is that slow. It hits the disk a bit, but do does ResolvePackageAssets. This could be related to nuget resolver slowness. ResolvePackageAssets cached all calls to the corresponding nuget API.

It is possible to do the same here. I will file a separate issue.

**Comment by nguerrera** [comment]
I guess in fairness, ResolvePackageAssets isn't resolving anything. I will compare to 2.0.

**Comment by nguerrera** [comment]
ReoslvePackageAssets is only 9ms for targeting 2.0.

ResolveFrameworkReferences is 43ms targeting 2.0 so not pay for play.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs [hunk 6]
```diff
@@ -46,7 +48,10 @@ protected override void ExecuteCore()
                 ITaskItem targetingPack;
                 if (!resolvedTargetingPacks.TryGetValue(frameworkReference.ItemSpec, out targetingPack))
                 {
-                    Log.LogError(Strings.UnknownFrameworkReference, frameworkReference.ItemSpec);
```
[the remark this answers is not in the document]
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs]: Do you mean the guideline "Do not have public members that can either throw or not throw exceptions based on some option."?

I don't consider our tasks public APIs, so I don't think this guideline applies.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs (hunk 3 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs]: I feel like it's OK to throw these types of errors when the invalid input is something that wouldn't be public if there was a concept of non-public items in MSBuild.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs (hunk 4 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: I think it was for parity with the runtime and apphost packs.  However, the asset resolution has changed since then so I can remove it.

**Comment by dsplaisted** [comment]
I've updated the `ResolveFrameworkReferences` task to short-circuit out if there are no `FrameworkReference` items, so it should be closer to pay-for-play now.

If we want the new tasks to be as fast as ResolvePackageAssets, we may need to cache them in the same way (or merge them into one task and cache it all in one file).

**Comment by nguerrera** [comment]
I filed a bug, #2908 for the perf.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.TargetingPackResolution.targets (hunk 5 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.TargetingPackResolution.targets]: Stage 0 needed to be updated, now done.

**Comment by dsplaisted** [comment]
Tests I want to add as part of this:

- [ ] Disable targeting pack download (EnableTargetingPackDownload)
- [ ] Runtime Framework version selection
  - [ ] RuntimeFrameworkVersion metadata on FrameworkReference item
  - [ ] RuntimeFrameworkVersion MSBuild property
  - [ ] TargetLatestRuntimePatch metadata on FrameworkReference
  - [ ] TargetLatestRuntimePatch MSBuild property
- [ ] Targeting pack version selection
- [ ] NETSDK1082 - No runtime pack available
- [ ] NETSDK1083 - RuntimeIdentifier not recognized (see It_fails_for_unsupported_rid and It_allows_unsupported_rid_with_override publish tests)

**Comment by dsplaisted** [comment]
@nguerrera @wli3 I'm done with the changes I want to make for this PR, if you would like to take a look and give an official sign-off.

🤞 that all the tests still pass 😄

**Comment by nguerrera** [comment]
I presume this should be retargeted to release/3.0.1xx branch.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+using System.Linq;
+using System.Text;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public class CheckForDuplicateFrameworkReferences : TaskBase
+    {
+        [Required]
+        public ITaskItem[] FrameworkReferenceItems { get; set; }
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs]: Why not name this `FrameworkReferences`. Items feels redundant with the `ITaskItem[]` type.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs [hunk 8]
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
...
+                return;
+            }
+
+            var knownFrameworkReferencesForTargetFramework = KnownFrameworkReferences.Select(item => new KnownFrameworkReference(item))
                 .Where(kfr => kfr.TargetFramework.Framework.Equals(TargetFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) &&
                               NormalizeVersion(kfr.TargetFramework.Version) == NormalizeVersion(new Version(TargetFrameworkVersion)))
-                .ToDictionary(kfr => kfr.Name);
+                .ToList();
+
+            var frameworkReferenceDict = FrameworkReferences.ToDictionary(fr => fr.ItemSpec);
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: nit: I'd pref `Map` over `Dict` abbreviation.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs [hunk 9]
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
...
-
-        [Output]
-        public string[] UnresolvedFrameworkReferences { get; set; }
+        public ITaskItem[] UnavailableRuntimePacks { get; set; }
 
         protected override void ExecuteCore()
         {
-            var knownFrameworkReferences = KnownFrameworkReferences.Select(item => new KnownFrameworkReference(item))
+            //  Perf optimization: If there are no FrameworkReference items, then don't do anything
+            //  (This means that if you don't have any direct framework references, you won't get any transitive ones either
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: I'm not sure this needs to be true. I would expect the transitive FrameworkReferences to be raised from assets files before we get here. Is it worth stating presumed implications on a feature that doesn't exist yet?

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        [Output]
+        public ITaskItem[] ItemsToAdd { get; set; }
+
+        [Output]
+        public ITaskItem[] ItemsToRemove { get; set; }
+
+
+        protected override void ExecuteCore()
+        {
+            var duplicateItems = FrameworkReferenceItems.GroupBy(i => i.ItemSpec, StringComparer.OrdinalIgnoreCase).Where(g => g.Count() > 1);
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs]: Let's put the early return on know framework references here too for consistency. Can't hurt to spare the allocations.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+        protected override void ExecuteCore()
+        {
+            var knownAppHostPacksForTargetFramework = KnownAppHostPacks
+                .Where(appHostPack =>
+                {
+                    var packTargetFramework = NuGetFramework.Parse(appHostPack.GetMetadata("TargetFramework"));
+                    return packTargetFramework.Framework.Equals(TargetFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) &&
+                        ResolveFrameworkReferences.NormalizeVersion(packTargetFramework.Version) ==
+                            ResolveFrameworkReferences.NormalizeVersion(new Version(TargetFrameworkVersion));
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs]: This normalization of right-hand-side can be hoisted.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs [hunk 12]
```diff
@@ -161,22 +161,23 @@ public void store_nativeonlyassets()
             storeDirectory.Should().OnlyHaveFiles(files_on_disk);
         }
 
-        [CoreMSBuildOnlyFact]
+        [CoreMSBuildOnlyFact(Skip = "https://github.com/dotnet/sdk/issues/2914")]
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs]: Is this test netcoreapp3.0 specifc?

Comment (nguerrera) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs]: @livarcocc Are you ok with disabling the store tests while we make a call. I'm a bit worried about taking on debt if we get the call to keep it.

File: src/Tests/Microsoft.NET.PerformanceTests/BuildPerf.cs [hunk 13]
```diff
@@ -71,6 +71,7 @@ public void BuildWebLarge(ProjectPerfOperation operation)
             var testDir = _testAssetsManager.CreateTestDirectory("WebLarge", identifier: operation.ToString());
             Console.WriteLine($"Mirroring {sourceProject} to {testDir}...");
             FolderSnapshot.MirrorFiles(sourceProject, testDir.Path);
+            TestContext.Current.WriteGlobalJson(testDir.Path);
```
Comment (nguerrera) [inline src/Tests/Microsoft.NET.PerformanceTests/BuildPerf.cs]: Why is this necessary?

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs (hunk 3 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs]: Should we put "_" in front of it if it wouldn't be public?

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs (hunk 9 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: This target runs before the assets file is read (it's the one that generates the list of packages that need to be downloaded).

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs (hunk 12 above)
Comment (dsplaisted) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs]: This one started failing for .NET Core 1.x.  I discussed disabling these tests with Livar, although I'm not sure I called out the fact that one of the failing ones wasn't for .NET Core 3.

File: src/Tests/Microsoft.NET.PerformanceTests/BuildPerf.cs (hunk 13 above)
Comment (dsplaisted) [reply src/Tests/Microsoft.NET.PerformanceTests/BuildPerf.cs]: So that if you run the perf tests with an `-sdkVersion` argument, it will respect that.  Most of the time the global.json will be written when the test directory is created, but in this case `MirrorFiles` wipes out anything in the target directory, so we have to explicitly write it again.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs (hunk 9 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: Ah, of course. The task name is a little ambiguous. I think of ResolveXReferences as something that turns abstract references into paths. Maybe we should rename it.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: I think the actual limitation in parentheses here is fine. I understand why it's needed now.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs (hunk 12 above)
Comment (nguerrera) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs]: It's a little concerning that we broke 1.x with this change. I can't think of a reason why that would happen. I'd prefer that we investigate in case we're breaking other aspects of exisitng projects.

File: src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs (hunk 7 above)
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs]: It was to resolve the singular/plural discrepancy between `FrameworkReference` as an MSBuild item name and `FrameworkReferences` as a task property :-)

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/CheckForDuplicateFrameworkReferences.cs]: Foos="@(Foo)" is common. I don't see issue.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs [hunk 14]
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
...
+
+                if (SelfContained &&
+                    !string.IsNullOrEmpty(RuntimeIdentifier) &&
+                    !string.IsNullOrEmpty(knownFrameworkReference.RuntimePackNamePatterns))
+                {
+                    foreach (var runtimePackNamePattern in knownFrameworkReference.RuntimePackNamePatterns.Split(';'))
                     {
-                        foreach (var runtimePackNamePattern in knownFrameworkReference.RuntimePackNamePatterns.Split(';'))
+                        string runtimePackRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
+                            new RuntimeGraphCache(this).GetRuntimeGraph(RuntimeGraphPath),
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: Can this be hoisted?

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: I guess it's cached, but still.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs [hunk 15]
```diff
@@ -55,168 +51,140 @@ public class ResolveFrameworkReferences : TaskBase
...
+                            RuntimeIdentifier,
+                            knownFrameworkReference.RuntimePackRuntimeIdentifiers.Split(';'),
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolveFrameworkReferences.cs]: This too.

**Review by nguerrera — COMMENTED** [review]
After realizing which task is ResolveFrameworkReferences, I'm very curious what takes so much time. I filed a separate bug already, so you can ignore the perf questions on this pass if you want.

**Comment by nguerrera** [comment]
I suspect the reading of the runtime graph.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs (hunk 12 above)
Comment (livarcocc) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToStoreAProjectWithDependencies.cs]: I wasn't aware that we were breaking 1.x projects as well. I thought the breaking was for netcoreapp3.0 projects.

**Comment by dsplaisted** [comment]
> I suspect the reading of the runtime graph.

@nguerrera I added a separate task to read the runtime graph, and that task took 12 ms while `ResolveFrameworkReferences` still took 24 ms.  So it doesn't seem to be just the loading of the runtime graph that's taking up the time.

**Comment by nguerrera** [comment]
See some of my comments on #2908. I'm also seeing a one-time init of FrameworkNameProvider being expensive. So far I haven't been able to profile accurately in a configuration that doesn't JIT nuget dlls. I think that's part of the answer here too.

**Comment by nguerrera** [comment]
I'll move some more thoughts on to #2908. Let's use that for perf discussion. I think with that issue filed, you can ignore my perf comments that came in late here with respect to following up on this PR feedback.
